Mutagen.Bethesda.Analyzers.Engine/Config/Topic/AnalyzerConfigReader.cs

```diff
@@ -0,0 +1,55 @@
+﻿using Microsoft.Extensions.Logging;
+using Mutagen.Bethesda.Analyzers.SDK.Topics;
+using Noggog;
+
+namespace Mutagen.Bethesda.Analyzers.Config.Topic;
+
+public class TopicConfigReader : IConfigReader<ITopicConfig>
+{
+    private const string DiagnosticGroup = "diagnostic";
+
+    private readonly ConfigReader<ITopicConfig> _reader;
+    private readonly ILogger<TopicConfigReader> _logger;
+
+    public TopicConfigReader(
+        ConfigReader<ITopicConfig> reader,
+        ILogger<TopicConfigReader> logger)
+    {
+        _logger = logger;
+        _reader = reader
```

Comment:
This pattern of doing the registration in the ctor feels like a code smell to me.   

I think it would be better if we perhaps had all the classes that were going to add to the reader logic instead implement something like `IRegisterToConfigReader` rather than being  IConfigReader themselves (aka, kill the decorator pattern aspect to it), and then did either:

# Factory Pattern
```
public class ConfigReaderFactory<TConfig>
{
   private readonly IRegisterToConfigReader<TConfig>[] _readerPlugins;
   public ConfigReaderFactory(IRegisterToConfigReader<TConfig>[] readerPlugins)
   {
       _readerPlugins = readerPlugins;
   } 

   public ConfigReader<TConfig> GetNewReader()
   {
       var ret = new ConfigReader<TConfig>();
       foreach (var plugin in _readerPlugins)
       {
           ret.Register(plugin);
       }
   }
}
```

# Self Mutation
```cs
public ConfigReader(
    IRegisterToConfigReader<TConfig>[] readerPlugins,
    ILogger<ConfigReader<TConfig>> logger)
{
    _logger = logger;
       foreach (var plugin in readerPlugins)
       {
           this.Register(plugin);
       }
}
```

This feels a bit better as at least the reader is doing the mutation itself, and has less external side effects

Reply:
I think the Factory code here would need to save the resulting reader object from each `Register()` call itself as it's returning a new object. So the reader is never mutated when you call `Register()`, it just creates an "extended" version of itself, adds the the passed function to the extended version's processing steps and returns the new object. The name "Register" isn't really telling that here though I think.

I'm not against changing the design of this implementation but I did like that you could simply inject the specific type of reader that you want like `AnalyzerConfigReader` to get all what you need. The factory has the extra step of calling the right method and it would need to know which config reader should use which processor units so we would put a lot of logic in there to handle that right? Similarly the self mutation approach would need some other class that is capsuling this class with a method to dynamically resolve the processor units it wants to inject into each reader via DI. At least as far as I can see?

Capsuling each processing step in an object could be nice from the perspective that you're not just dealing with raw Funcs but that could probably also be added independently if we want that. Though of course it kind of inflates the codebase when there is a class each, and they would need to be resolved via DI so they can also use things like loggers. Not quite sure, seems like everything has it's upsides and downsides 😄
What do you think are the biggest problems with the ctor registrations in the current decorator pattern with builder-like registrations that at least work with immutable readers?

Reply:
> I think I'll go with the self mutation approach!

The reason I like it is that the container helps us out a bit.  We toss in all the IConfigReaderProcessorUnit ingredients, and it pulls them all together to give to ConfigReader

This is a similar pattern to how the analyzers are wired together, where `ContextualDriver` requests all the `IContextualAnalyzer[] contextualAnalyzers` from the container.

> What do you think are the biggest problems with the ctor registrations in the current decorator pattern with builder-like registrations that at least work with immutable readers?

My current reasons are:
- I like doing functional programming where I can.  Just a learned preferred practice.  Unexpected mutations introduce edge cases
- There's nothing stopping the system from instantiating a latecomer that does a registration after the fact, at which point the parent config reader's logic suddenly changes due to the registration that suddenly occurred.    Probably wont happen.. but requesting all the `IConfigReaderProcessorUnit` ahead of time is just much clearer.  We know all the pieces that will make up the reader at instantiation time, and here they all are.   It just closes off whole categories of bugs.
- Confusion on the decoration order.   Whos the top decoration?  Whos the nested decoration?  Does it matter?  Can it be strung up wrong if done in the wrong order?   Lots to think about.   Vs injecting in `IConfigReaderProcessorUnit[]` is more easy to think aobut.   We're making the config reader, and we have 8 plugins right here for it to use.  Easy